File: src/commands/claude.ts
import * as child_process from 'child_process';
import { Credentials } from '../credentials';

export interface ClaudeDeps {
  credentials?: Credentials;
  env?: Record<string, string | undefined>;
  spawn?: typeof child_process.spawn;
}

export async function claudeCommand(args: string[], deps?: ClaudeDeps): Promise<void> {
  const credentials = deps?.credentials ?? new Credentials();
  const parentEnv = deps?.env ?? process.env;
  const spawnFn = deps?.spawn ?? child_process.spawn;

  const creds = credentials.read();
  if (!creds) {
    process.stderr.write('Not logged in. Run `monocle login --tenant <domain>` first.\n');
    process.exit(1);
    return;
  }

  // Resolve router URL (base URL for the Chat Proxy)
  let routerUrl: string;
  if (creds.router_url) {
    routerUrl = creds.router_url;
  } else {
    const isLocal = creds.tenant_domain.startsWith('localhost') || creds.tenant_domain.startsWith('127.0.0.1');
    routerUrl = `${isLocal ? 'http' : 'https'}://${creds.tenant_domain}`;
  }

  // Inline settings scoped to this child only — avoids mutating ~/.claude/settings.json.
  // `apiKeyHelper` keeps tokens fresh across long sessions by re-invoking `monocle token`.
  // `model` defaults to the 1M-context Sonnet to avoid surprise Opus costs; a user `--model` flag
  // (forwarded after --settings) still wins, since the CLI flag outranks the settings field.
  const inlineSettings = JSON.stringify({
    apiKeyHelper: 'monocle token',
    model: 'sonnet[1m]',
  });

  // Build child env — strip conflicting vars, inject base URL.
  const childEnv: Record<string, string> = {};
  for (const [key, value] of Object.entries(parentEnv)) {
    if (value !== undefined && key !== 'ANTHROPIC_API_KEY' && key !== 'ANTHROPIC_AUTH_TOKEN') {
      childEnv[key] = value;
    }
  }
  childEnv.ANTHROPIC_BASE_URL = routerUrl;

  const child = spawnFn('claude', ['--settings', inlineSettings, ...args], {
    env: childEnv,
    stdio: 'inherit',
  });

  let exited = false;

  child.on('close', (code) => {
    if (!exited) { exited = true; process.exit(code ?? 0); }
  });

  child.on('error', (err: NodeJS.ErrnoException) => {
    if (!exited) {
      exited = true;
      if (err.code === 'ENOENT') {
        process.stderr.write('Error: `claude` command not found. Is Claude Code installed?\n');
      } else {
        process.stderr.write(`Error: ${err.message}\n`);
      }
      process.exit(1);
    }
  });
}

File: src/commands/claude.rs
use std::process::Command;

use serde_json::json;

use crate::auth::router_url_for;
use crate::credentials::Credentials;
use crate::origin::{MONOCLE_ORIGIN, ORIGIN_HEADER_NAME};

/// Launch Claude Code through Monocle: inline settings, conflicting env vars
/// stripped, base URL injected, usage attributed to the "cli" surface.
pub fn claude_command(creds: &Credentials, args: &[String]) {
    let creds = match creds.read() {
        Some(c) => c,
        None => {
            eprintln!("Not logged in. Run `monocle login --tenant <domain>` first.");
            std::process::exit(1);
        }
    };

    let router_url = router_url_for(&creds);

    // Inline settings scoped to this child only — avoids mutating
    // ~/.claude/settings.json. `apiKeyHelper` keeps tokens fresh across long
    // sessions; `model` defaults to the 1M-context Sonnet to avoid surprise Opus
    // costs (a user `--model` flag still wins, since the CLI flag outranks the
    // settings field).
    let inline_settings = json!({
        "apiKeyHelper": "monocle token",
        "model": "sonnet[1m]",
    })
    .to_string();

    // Merge the origin header into ANTHROPIC_CUSTOM_HEADERS (newline-separated
    // `Name: Value` lines) so we don't clobber any the user already set.
    let origin_line = format!("{ORIGIN_HEADER_NAME}: {MONOCLE_ORIGIN}");
    let custom_headers = match std::env::var("ANTHROPIC_CUSTOM_HEADERS") {
        Ok(existing) if !existing.is_empty() => format!("{existing}\n{origin_line}"),
        _ => origin_line,
    };

    let mut cmd = Command::new("claude");
    cmd.arg("--settings")
        .arg(&inline_settings)
        .args(args)
        .env_remove("ANTHROPIC_API_KEY")
        .env_remove("ANTHROPIC_AUTH_TOKEN")
        .env("ANTHROPIC_BASE_URL", &router_url)
        .env("ANTHROPIC_CUSTOM_HEADERS", &custom_headers);

    run(cmd);
}

#[cfg(unix)]
fn run(mut cmd: Command) {
    use std::os::unix::process::CommandExt;
    // exec replaces this process — stdio, TTY, signals, and exit code all pass
    // through transparently. Only returns on failure to launch.
    let err = cmd.exec();
    report_spawn_error(err);
}

#[cfg(not(unix))]
fn run(mut cmd: Command) {
    match cmd.status() {
        Ok(status) => std::process::exit(status.code().unwrap_or(0)),
        Err(err) => report_spawn_error(err),
    }
}

fn report_spawn_error(err: std::io::Error) -> ! {
    if err.kind() == std::io::ErrorKind::NotFound {
        eprintln!("Error: `claude` command not found. Is Claude Code installed?");
    } else {
        eprintln!("Error: {err}");
    }
    std::process::exit(1);
}
